src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,168 @@
+        Entry<T> newEntry = new Entry<>(value, null, last);
+        if (size == 0) {
+            first = newEntry;
+        } else {
+            newEntry.prev.next = newEntry;
+        }
+        last = newEntry;
+        ++size;
+    }
+
+    private void addByIndex(T value, int index) {
+        if (index == size) {
+            addInTheEnd(value);
+        } else {
+            Entry<T> entryAtIndex = findByIndex(index);
+            Entry<T> oldHead = last;
+            if (entryAtIndex.prev == null) {
+                first = new Entry<>(value, entryAtIndex, null);
+            } else {
+                Entry<T> newNode = new Entry<T>(value, entryAtIndex, entryAtIndex.prev);
+                newNode.prev.next = newNode;
+                entryAtIndex.prev = newNode;
+            }
+            size++;
+        }
+    }
+
+    private Entry<T> findByIndex(int index) {
+        Entry<T> entryToFind;
+        if (size / 2 >= index) {
+            entryToFind = first;
+            for (int i = 0; i < index; i++) {
+                entryToFind = entryToFind.next;
+            }
+        } else {
+            entryToFind = last;
+            for (int i = 0; i < size - index - 1; i++) {
+                entryToFind = entryToFind.prev;
+            }
+        }
+        return entryToFind;
+    }
+
     @Override
     public boolean add(T value) {
-        return false;
+        addInTheEnd(value);
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkExeption(index);
+        addByIndex(value, index);
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        if (list.size() == 0) {
+            return false;
+        } else {
+            for (int i = 0; i < list.size(); i++) {
+                add(list.get(i));
+            }
+            return true;
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkExeptionFilledElement(index);
+        Entry<T> element = findByIndex(index);
+        return element.element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkExeptionFilledElement(index);
+        Entry<T> oldElement = findByIndex(index);
+        T oldValue = oldElement.element;
+        oldElement.element = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkExeptionFilledElement(index);
+        Entry<T> removeElement = findByIndex(index);
+        Entry<T> previousToRemove = removeElement.prev;
+        Entry<T> nextToRemove = removeElement.next;
+        if (previousToRemove == null) {
+            first = nextToRemove;
+        } else {
+            previousToRemove.next = nextToRemove;
+        }
+        if (nextToRemove == null) {
+            last = previousToRemove;
+        } else {
+            nextToRemove.prev = previousToRemove;
+        }
+        size--;
```

Comment:
let's move this code at the separate method `unlink(Node<T> nodeToRemove)` and then we can call it in the remove(T t);